examples/hitl/rearrange_v2/ui.py

```diff
@@ -0,0 +1,448 @@
+                self._sim, object_id, self._link_id_to_ao_map
+            )
+            link_node = ao.get_link_scene_node(link_index)
+            aabb = link_node.cumulative_bb
+            reachable = self._is_within_reach(link_node.translation)
+            color = COLOR_VALID if reachable else COLOR_INVALID
+            self._draw_aabb(aabb, link_node.transformation, color)
+
+    def _draw_hovered_pickable(self) -> None:
+        """Highlight the hovered pickable object."""
+        if not self._hover_selection.selected or self._is_holding_object():
+            return
+
+        object_id = self._hover_selection.object_id
+        if not self._is_object_pickable(object_id):
+            return
+
+        managed_object = sim_utilities.get_obj_from_id(
+            self._sim, object_id, self._link_id_to_ao_map
+        )
+        translation = managed_object.translation
+        reachable = self._is_within_reach(translation)
+        color = COLOR_VALID if reachable else COLOR_INVALID
+        aabb = managed_object.collision_shape_aabb
+        self._draw_aabb(aabb, managed_object.transformation, color)
+
+    def _draw_goals(self) -> None:
+        """Draw goal object-receptacle pairs."""
+        # TODO: Cache
+        for i in range(len(self._object_receptacle_pairs)):
+            rigid_ids = self._object_receptacle_pairs[i][0]
+            receptacle_ids = self._object_receptacle_pairs[i][1]
+            goal_pair_color = COLOR_GOALS[i % len(COLOR_GOALS)]
+            for rigid_id in rigid_ids:
+                if self._hover_selection.object_id == rigid_id:
+                    continue
+                managed_object = sim_utilities.get_obj_from_id(
+                    self._sim, rigid_id, self._link_id_to_ao_map
+                )
+                translation = managed_object.translation
+                self._gui_drawer.draw_circle(
+                    translation=translation,
+                    radius=0.25,
+                    color=goal_pair_color,
+                    billboard=True,
+                    destination_mask=self._dest_mask,
+                )
+            for receptacle_id in receptacle_ids:
+                managed_object = sim_utilities.get_obj_from_id(
+                    self._sim, receptacle_id, self._link_id_to_ao_map
+                )
+                aabb = None
+                if hasattr(managed_object, "collision_shape_aabb"):
+                    aabb = managed_object.collision_shape_aabb
+                else:
+                    link_index = self._get_link_index(receptacle_id)
+                    if link_index is not None:
+                        link_node = managed_object.get_link_scene_node(
+                            link_index
+                        )
+                        aabb = link_node.cumulative_bb
+                if aabb is not None:
+                    self._draw_aabb(
+                        aabb, managed_object.transformation, goal_pair_color
+                    )
```

Comment:
Here's a suggestion to limit the map/dict lookups inherent in dot-dereferencing (such as with _self_ ) in python by making local references before the loop starts to both member variables and methods that are called often within the nested loops.  

Not sure how large the object_receptacle_pairs list is, nor the subsequent sublists of rigid_ids and receptacle_ids, but if these are expected to possibly be large, these changes might make a noticeable difference in draw speed.

-shrug- ymmv :) 

```suggestion
        sim = self._sim
        obj_receptacle_pairs = self._object_receptacle_pairs
        link_id_to_ao_map = self._link_id_to_ao_map
        dest_mask = self._dest_mask
        # can do this with methods as well - note the absence of parenthesis
        get_obj_from_id = sim_utilities.get_obj_from_id
        get_link_index = self._get_link_index
        draw_gui_circle = self._gui_drawer.draw_circle
        draw_gui_aabb = self._draw_aabb
        
        for i in range(len(obj_receptacle_pairs)):
            rigid_ids = obj_receptacle_pairs[i][0]
            receptacle_ids = obj_receptacle_pairs[i][1]
            goal_pair_color = COLOR_GOALS[i % len(COLOR_GOALS)]
            for rigid_id in rigid_ids:
                if self._hover_selection.object_id == rigid_id:
                    continue
                managed_object = get_obj_from_id(
                    sim, rigid_id, link_id_to_ao_map
                )
                translation = managed_object.translation
                draw_gui_circle(
                    translation=translation,
                    radius=0.25,
                    color=goal_pair_color,
                    billboard=True,
                    destination_mask=dest_mask,
                )
            for receptacle_id in receptacle_ids:
                managed_object = get_obj_from_id(
                    sim, receptacle_id, link_id_to_ao_map
                )
                aabb = None
                if hasattr(managed_object, "collision_shape_aabb"):
                    aabb = managed_object.collision_shape_aabb
                else:
                    link_index = get_link_index(receptacle_id)
                    if link_index is not None:
                        link_node = managed_object.get_link_scene_node(
                            link_index
                        )
                        aabb = link_node.cumulative_bb
                if aabb is not None:
                    draw_gui_aabb(
                        aabb, managed_object.transformation, goal_pair_color
                    )
```